In [1]:

!pip install tensorflow-gpu==2.0.0-rc1

In [0]:
from tensorflow.keras.models import Sequential 

In [0]:

import warnings
warnings.filterwarnings("ignore")

from tensorflow.keras.datasets import mnist

In [0]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

input_dim = 784  # 28*28
output_dim = nb_classes = 10
batch_size = 128
nb_epoch = 20

X_train = X_train.reshape(60000, input_dim)
X_test = X_test.reshape(10000, input_dim)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255

1. In this task, you'll implement several ANN models with different activation functions. Specifically:

Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using tanh activation function for each layer.
Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using sigmoid activation function for each layer.
Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using ReLU activation function for each layer.
Compare the result of each model with each other. Which activation function did perform better?

1. Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using tanh activation function for each layer.

In [0]:
from tensorflow.keras.utils import to_categorical

Y_train = to_categorical(y_train, nb_classes)
Y_test = to_categorical(y_test, nb_classes)

In [6]:
X_train[0].shape

(784,)

In [0]:
from tensorflow.keras.models import Sequential 
from tensorflow.keras.layers import Dense

model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="tanh"))
# our second dense layer
model.add(Dense(64, activation="tanh"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

In [8]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 128)               100480    
_________________________________________________________________
dense_1 (Dense)              (None, 64)                8256      
_________________________________________________________________
dense_2 (Dense)              (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [0]:
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

In [10]:
# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 2s 35us/sample - loss: 1.0283 - accuracy: 0.7571
Epoch 2/20
60000/60000 [==============================] - 2s 28us/sample - loss: 0.5176 - accuracy: 0.8724
Epoch 3/20
60000/60000 [==============================] - 2s 28us/sample - loss: 0.4188 - accuracy: 0.8900
Epoch 4/20
60000/60000 [==============================] - 2s 28us/sample - loss: 0.3717 - accuracy: 0.8994
Epoch 5/20
60000/60000 [==============================] - 2s 28us/sample - loss: 0.3424 - accuracy: 0.9061
Epoch 6/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.3218 - accuracy: 0.9103
Epoch 7/20
60000/60000 [==============================] - 2s 28us/sample - loss: 0.3058 - accuracy: 0.9146
Epoch 8/20
60000/60000 [==============================] - 2s 28us/sample - loss: 0.2926 - accuracy: 0.9175
Epoch 9/20
60000/60000 [==============================] - 2s 28us/sample - loss: 0.2815 - accuracy: 0.9203
Epoch 10/20
60

In [11]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.2022271488547325
Test accuracy: 0.9408


2. Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using sigmoid activation function for each layer.

In [0]:
from tensorflow.keras.models import Sequential 
from tensorflow.keras.layers import Dense

model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="sigmoid"))
# our second dense layer
model.add(Dense(64, activation="sigmoid"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

In [13]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_3 (Dense)              (None, 128)               100480    
_________________________________________________________________
dense_4 (Dense)              (None, 64)                8256      
_________________________________________________________________
dense_5 (Dense)              (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [0]:
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

In [15]:
# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 2s 33us/sample - loss: 2.2897 - accuracy: 0.2294
Epoch 2/20
60000/60000 [==============================] - 2s 28us/sample - loss: 2.2163 - accuracy: 0.3826
Epoch 3/20
60000/60000 [==============================] - 2s 29us/sample - loss: 2.1434 - accuracy: 0.4809
Epoch 4/20
60000/60000 [==============================] - 2s 28us/sample - loss: 2.0382 - accuracy: 0.5421
Epoch 5/20
60000/60000 [==============================] - 2s 28us/sample - loss: 1.8911 - accuracy: 0.5745
Epoch 6/20
60000/60000 [==============================] - 2s 28us/sample - loss: 1.7121 - accuracy: 0.6066
Epoch 7/20
60000/60000 [==============================] - 2s 28us/sample - loss: 1.5288 - accuracy: 0.6450
Epoch 8/20
60000/60000 [==============================] - 2s 28us/sample - loss: 1.3635 - accuracy: 0.6835
Epoch 9/20
60000/60000 [==============================] - 2s 27us/sample - loss: 1.2221 - accuracy: 0.7140
Epoch 10/20
60

In [16]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.5614777091026306
Test accuracy: 0.8534


3. Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using ReLU activation function for each layer.

In [0]:
from tensorflow.keras.models import Sequential 
from tensorflow.keras.layers import Dense

model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="relu"))
# our second dense layer
model.add(Dense(64, activation="relu"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

In [18]:
model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_6 (Dense)              (None, 128)               100480    
_________________________________________________________________
dense_7 (Dense)              (None, 64)                8256      
_________________________________________________________________
dense_8 (Dense)              (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [0]:
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

In [20]:
# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 2s 34us/sample - loss: 1.2395 - accuracy: 0.6898
Epoch 2/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.5079 - accuracy: 0.8671
Epoch 3/20
60000/60000 [==============================] - 2s 28us/sample - loss: 0.3965 - accuracy: 0.8901
Epoch 4/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.3508 - accuracy: 0.9008
Epoch 5/20
60000/60000 [==============================] - 2s 28us/sample - loss: 0.3231 - accuracy: 0.9078
Epoch 6/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.3029 - accuracy: 0.9133
Epoch 7/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.2865 - accuracy: 0.9181
Epoch 8/20
60000/60000 [==============================] - 2s 28us/sample - loss: 0.2726 - accuracy: 0.9217
Epoch 9/20
60000/60000 [==============================] - 2s 28us/sample - loss: 0.2601 - accuracy: 0.9252
Epoch 10/20
60

In [21]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.17382131612598897
Test accuracy: 0.9491


RESULTS: Using tanh, accuracy =0.941
      Using sigmoid, accuracy =0.853
      Using relu, accuracy =0.949

Conclusion: relu seemed to perform best

In this task, you'll implement the ANN models specified below with the hinge loss function as their loss functions:

Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using tanh activation function for each layer.
Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using sigmoid activation function for each layer.
Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using ReLU activation function for each layer.
Compare the result of each model with the result of the same model from the previous task. Which loss function did perform better?

1. Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using tanh activation function for each layer.

In [23]:
model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="tanh"))
# our second dense layer
model.add(Dense(64, activation="tanh"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

model.compile(optimizer='sgd', loss='categorical_hinge',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
60000/60000 [==============================] - 2s 37us/sample - loss: 0.9795 - accuracy: 0.4173
Epoch 2/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.7709 - accuracy: 0.6656
Epoch 3/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.6058 - accuracy: 0.7720
Epoch 4/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.4920 - accuracy: 0.8348
Epoch 5/20
60000/60000 [==============================] - 2s 30us/sample - loss: 0.4102 - accuracy: 0.8607
Epoch 6/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.3565 - accuracy: 0.8740
Epoch 7/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.3210 - accuracy: 0.8821
Epoch 8/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.2958 - accuracy: 0.8889
Epoch 9/20
60000/60000 [===

In [24]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.18611867011189462
Test accuracy: 0.9212


2. Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using sigmoid activation function for each layer.

In [25]:
model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="sigmoid"))
# our second dense layer
model.add(Dense(64, activation="sigmoid"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

model.compile(optimizer='sgd', loss='categorical_hinge',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 2s 34us/sample - loss: 1.0351 - accuracy: 0.1037
Epoch 2/20
60000/60000 [==============================] - 2s 29us/sample - loss: 1.0096 - accuracy: 0.1299
Epoch 3/20
60000/60000 [==============================] - 2s 29us/sample - loss: 1.0027 - accuracy: 0.2025
Epoch 4/20
60000/60000 [==============================] - 2s 28us/sample - loss: 1.0022 - accuracy: 0.2525
Epoch 5/20
60000/60000 [==============================] - 2s 29us/sample - loss: 1.0017 - accuracy: 0.2967
Epoch 6/20
60000/60000 [==============================] - 2s 29us/sample - loss: 1.0012 - accuracy: 0.3432
Epoch 7/20
60000/60000 [==============================] - 2s 28us/sample - loss: 1.0008 - accuracy: 0.3895
Epoch 8/20
60000/60000 [==============================] - 2s 30us/sample - loss: 1.0005 - accuracy: 0.4326
Epoch 9/20
60000/60000 [==============================] - 2s 29us/sample - loss: 1.0001 - accuracy: 0.4687
Epoch 10/20
60

In [26]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.9967496450424195
Test accuracy: 0.6657


3. Implement a three layer ANN model with 128, 64 and 10 neurons in the layers using ReLU activation function for each layer.

In [28]:
model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="relu"))
# our second dense layer
model.add(Dense(64, activation="relu"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

model.compile(optimizer='sgd', loss='categorical_hinge',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=batch_size, epochs=20, verbose=1)

Train on 60000 samples
Epoch 1/20
60000/60000 [==============================] - 2s 35us/sample - loss: 1.0176 - accuracy: 0.2333
Epoch 2/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.9769 - accuracy: 0.5166
Epoch 3/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.8930 - accuracy: 0.6536
Epoch 4/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.7305 - accuracy: 0.7349
Epoch 5/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.5358 - accuracy: 0.8064
Epoch 6/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.4070 - accuracy: 0.8620
Epoch 7/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.3318 - accuracy: 0.8799
Epoch 8/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.2892 - accuracy: 0.8893
Epoch 9/20
60000/60000 [==============================] - 2s 29us/sample - loss: 0.2623 - accuracy: 0.8952
Epoch 10/20
60

In [29]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.163042075496912
Test accuracy: 0.9279


RESULTS: Using tanh, accuracy =0.921
      Using sigmoid, accuracy =0.666
      Using relu, accuracy =0.928

Conclusion: relu seemed to perform best